File: benchmarks/js/async/scheduler.js
// Generator scheduler benchmark
// Create 100 generators, each counting from 0 to N/100, round-robin schedule

var print = typeof console !== 'undefined' ? console.log.bind(console) : globalThis.print;
var DEFAULT = 100000;
var N = parseInt(typeof scriptArgs !== 'undefined' ? scriptArgs[1] : (typeof process !== 'undefined' ? process.argv[2] : '0')) || DEFAULT;

var numGenerators = 100;
var perGenerator = Math.floor(N / numGenerators);

function* worker(limit) {
    for (var i = 0; i < limit; i++) {
        yield;
    }
}

// Create generators
var generators = new Array(numGenerators);
for (var i = 0; i < numGenerators; i++) {
    generators[i] = worker(perGenerator);
}

// Round-robin scheduler
var totalYields = 0;
var alive = numGenerators;
while (alive > 0) {
    for (var i = 0; i < numGenerators; i++) {
        var gen = generators[i];
        if (gen) {
            var result = gen.next();
            if (result.done) {
                generators[i] = null;
                alive--;
            } else {
                totalYields++;
            }
        }
    }
}

print(totalYields);

File: benchmarks/lua/coroutine/scheduler.lua
-- Coroutine scheduler benchmark
-- Create 100 coroutines, each counting from 0 to N/100, round-robin schedule

local N = tonumber(arg and arg[1]) or 5000000
-- collectgarbage("collect")

local num_coroutines = 100
local per_coroutine = math.floor(N / num_coroutines)

local function worker(limit)
    for i = 0, limit - 1 do
        coroutine.yield()
    end
end

-- Create coroutines
local coroutines = {}
for i = 1, num_coroutines do
    coroutines[i] = coroutine.create(function() worker(per_coroutine) end)
end

-- Round-robin scheduler
local total_yields = 0
local alive = num_coroutines
while alive > 0 do
    for i = 1, num_coroutines do
        local co = coroutines[i]
        if co then
            local ok, _ = coroutine.resume(co)
            if coroutine.status(co) == "dead" then
                coroutines[i] = nil
                alive = alive - 1
            else
                total_yields = total_yields + 1
            end
        end
    end
end

print(total_yields)
